From 971965f08b63834c10e50d07d70cd6db31cabdac Mon Sep 17 00:00:00 2001
From: Carlos Leija <cileija@ti.com>
Date: Wed, 9 Mar 2011 15:07:41 -0600
Subject: OMAP4: Set PL310 POR register in kernel

Added functionality to set PL310 in kernel and not have dependency
on bootloader code

Change-Id: I1dcc502640479eb1ac1c07e9f047c56034f0bb8a
Signed-off-by: Carlos Leija <cileija@ti.com>
---
 arch/arm/mach-omap2/include/mach/omap4-common.h |    2 ++
 arch/arm/mach-omap2/omap4-common.c              |    9 ++++++++-
 2 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-omap2/include/mach/omap4-common.h b/arch/arm/mach-omap2/include/mach/omap4-common.h
index 7cdc03c..f8ca41f 100644
--- a/arch/arm/mach-omap2/include/mach/omap4-common.h
+++ b/arch/arm/mach-omap2/include/mach/omap4-common.h
@@ -44,8 +44,10 @@
 /*
  * Secure HAL, PPA services available
  */
+#define PPA_SERVICE_PL310_POR		0x23
 #define PPA_SERVICE_NS_SMP		0x25
 
+#define PL310_POR			5
 /*
  * Secure HAL API flags
  */
diff --git a/arch/arm/mach-omap2/omap4-common.c b/arch/arm/mach-omap2/omap4-common.c
index 34b9694..44df3f5 100644
--- a/arch/arm/mach-omap2/omap4-common.c
+++ b/arch/arm/mach-omap2/omap4-common.c
@@ -219,8 +219,15 @@ static int __init omap_l2_cache_init(void)
 	l2cache_base = ioremap(OMAP44XX_L2CACHE_BASE, SZ_4K);
 	BUG_ON(!l2cache_base);
 
-	if (omap_rev() != OMAP4430_REV_ES1_0)
+	if (omap_rev() != OMAP4430_REV_ES1_0) {
+		/* Set POR through PPA service only in EMU/HS devices */
+		if (omap_type() != OMAP2_DEVICE_TYPE_GP)
+			omap4_secure_dispatcher(
+				PPA_SERVICE_PL310_POR, 0x7, 1,
+				PL310_POR, 0, 0, 0);
+
 		omap_smc1(0x109, OMAP4_L2X0_AUXCTL_VALUE);
+	}
 
 	/* Enable PL310 L2 Cache controller */
 	omap_smc1(0x102, 0x1);
-- 
1.6.0.4

